### ISSUE TITLE
[Issue]: An issue with vary_coeff in UniPC sampler

### ISSUE THREAD

**Issue by mart-hill** [vladmandic/automatic#663]
### Issue Description

At first I noticed, that I still have hires fix on for 1024x1024 image generation with "legacy" hires upscaling ON, ControlNet tile preprocessor and two LoRAs inside Additional Networks extension, so the error Tiled VAE extension threw at me was "viable". I'm mentioning this, so it could be recreated. 🙂 
When I turned off hires fix, and had Multidiffusion "overtake" the resolution with Multiple Diffusers option, UniPC normally would generate an image, but this time, as I have "vary_coeff" selected in sampler's settings, I've got this error (as I noticed, it doesn't matter, if I use Multidiffusion/Tiled VAE or not), this error will happen if I have "vary_coeff" selected in Settings:
```
Loading model from cache: control_v11u_sd15_tile [1f041471]
Loading preprocessor: tile_resample
Pixel Perfect Mode Enabled.
resize_mode = ResizeMode.OUTER_FIT
raw_H = 1500
raw_W = 1500
target_H = 1024
target_W = 1024
estimation = 1024.0
preprocessor resolution = 1024
Initializing ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:-- 0:00:01
gradio call: RuntimeError
╭─────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────╮
│ X:\AI\automatic\modules\call_queue.py:58 in f                                                                    │
│                                                                                                                  │
│    57 │   │   │   │   pr.enable()                                                                                │
│ ❱  58 │   │   │   res = list(func(*args, **kwargs))                                                              │
│    59 │   │   │   if shared.cmd_opts.profile:                                                                    │
│                                                                                                                  │
│ X:\AI\automatic\modules\call_queue.py:38 in f                                                                    │
│                                                                                                                  │
│    37 │   │   │   try:                                                                                           │
│ ❱  38 │   │   │   │   res = func(*args, **kwargs)                                                                │
│    39 │   │   │   finally:                                                                                       │
│                                                                                                                  │
│                                             ... 11 frames hidden ...                                             │
│                                                                                                                  │
│ X:\AI\automatic\modules\models\diffusion\uni_pc\uni_pc.py:523 in multistep_uni_pc_update                         │
│                                                                                                                  │
│   522 │   │   │   assert self.variant == 'vary_coeff'                                                            │
│ ❱ 523 │   │   │   return self.multistep_uni_pc_vary_update(x, model_prev_list, t_prev_list, t,                   │
│   524                                                                                                            │
│                                                                                                                  │
│ X:\AI\automatic\modules\models\diffusion\uni_pc\uni_pc.py:587 in multistep_uni_pc_vary_update                    │
│                                                                                                                  │
│   586 │   │   │   x_t_ = (                                                                                       │
│ ❱ 587 │   │   │   │   sigma_t / sigma_prev_0 * x                                                                 │
│   588 │   │   │   │   - alpha_t * h_phi_1 * model_prev_0                                                         │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: The size of tensor a (2) must match the size of tensor b (128) at non-singleton dimension 3
```

A little side mention (but it's probably gradio's fault) - when I went back to bh1 option, saved, and pressed Generate, the UI hung. It was the usual lock-up without generating a picture and without any error message (this race-condition bug). Since it happened while I was setting up things in UniPC sampler section of settings, I'm mentioning it here - sorry about that. 🙂 

### Version Platform Description

Windows 22H2, RTX 3090, 64GB of RAM

```
04:35:16-611788 INFO     Python 3.10.9 on Windows
04:35:17-647788 INFO     Version: 4dc59419 Sun Apr 30 21:01:49 2023 -0400
04:35:18-224787 INFO     Setting environment tuning
04:35:18-244787 INFO     nVidia CUDA toolkit detected
04:35:21-028287 INFO     Torch 2.0.0+cu118
04:35:21-053288 INFO     Torch backend: nVidia CUDA 11.8 cuDNN 8700
04:35:21-055288 INFO     Torch detected GPU: NVIDIA GeForce RTX 3090 VRAM 24575 Arch (8, 6) Cores 82
```

**Comment by vladmandic** [comment]
@mart-hill you have a special gift of finding those issues! :)

looping in @cool-bigdogs-tshirt as he did unipc patch to enable img2img workflows.

**Comment by cool-bigdogs-tshirt** [comment]
will be able to look more closely in ~8 hours. if you have time before then (since you can already reproduce), mind popping these prints in `automatic\modules\models\diffusion\uni_pc\uni_pc.py:587`? not sure which of the two tensors out of those six in the statement have the wrong shape:

```
        if self.predict_x0:
            print(f' sigma_t: {sigma_t.shape})
            print(f' sigma_prev_0: {sigma_prev_0.shape})
            print(f' x: {x.shape})
            print(f' alpha_t: {alpha_t.shape})
            print(f' h_phi_1: {h_phi_1.shape})
            print(f' model_prev_0: {model_prev_0.shape})
            x_t_ = (
                sigma_t / sigma_prev_0 * x
                - alpha_t * h_phi_1 * model_prev_0
            )
```

also uhhh i'm still just guessing at all the actual ML stuff here, but knowing which data is in the wrong shape should help at least getting the code to run again.

**Comment by cool-bigdogs-tshirt** [comment]
and actually, that might spam a billion lines if it only breaks later on... maybe something like this:
```
        if self.predict_x0:
            _a = sigma_t / sigma_prev_0
            _b = _a * x
            
            _c = alpha_t * h_phi_1
            _d = _c * model_prev_0

            x_t_ = _b - _d
```

**Comment by cool-bigdogs-tshirt** [comment]
actually, looks like maybe nobody's done batching with vary_coeff before? the fix for batching seems to have just been sprinkling `expand_dims` all over the code :(

here it is in the `bh` function, could probably just copy paste that: https://github.com/vladmandic/automatic/blob/master/modules/models/diffusion/uni_pc/uni_pc.py#L706-L709

i wonder how bad this is for performance :thinking:

**Comment by mart-hill** [comment]
I did have previously some, like, 4 images generated as once with UniPC, with vary_coeff, but that was before this error popped up. 🙂

**Comment by mart-hill** [comment]
I can generate images with vary_coeff on 22da90d4 commit (and one earlier back, too), but the generation of the images is suddenly way slower on all samplers, it's about 1.4-2.5 it/s on 512x512 image with RTX 3090. I did `webui --upgrade` though.

The quality of UniPC is also really bad:
This is UniPC with this settings:
![image](https://user-images.githubusercontent.com/55754730/235491509-9041f85b-f60c-4011-adfe-05d5e09e6944.png)

![00107-20230501185108_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235491440-2656493f-c539-4d22-8040-01ec8db4b108.png)

This is Euler a with NO negative prompt (UniPC image had to have "bad quality", or the image would be WAY worse).
![00103-20230501184443_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235491443-70967e89-6964-499b-a9fe-5716a3024e42.png)

**Comment by cool-bigdogs-tshirt** [comment]
what other unipc settings do you have? i've been sticking with 'skip_type' of 'uniform', I think other schedulers might make it look worse?

it's very strange that it works on that commit since it seems like the code shouldn't work (based on the `expand_dims` function being in the `bh` method but not where you're having a problem), and hasn't been changed recently.

**Comment by cool-bigdogs-tshirt** [comment]
d'oh i just saw your screenshot. i will try some other skip types, my code generating the timesteps for denoising doesn't take the different schedules for unipc into account and i bet that's why it looks worse.

**Comment by mart-hill** [comment]
> @mart-hill performance is due to new code added in `a1111-sd-webui-lycoris`, author is investigating. new code attempted to fix memory leak, but introduced a massive performance hit.

Ohh. It is MASSIVE. It' even like 5x slower sometimes.

**Comment by mart-hill** [comment]
> what other unipc settings do you have? i've been sticking with 'skip_type' of 'uniform', I think other schedulers might make it look worse?
> 
> it's very strange that it works on that commit since it seems like the code shouldn't work (based on the `expand_dims` function being in the `bh` method but not where you're having a problem), and hasn't been changed recently.

Should I show classic SD15 512x512 outputs for various UniPC variants?

**Comment by cool-bigdogs-tshirt** [comment]
'uniform' should give you good hires results for now.

i learned all this from messing with the code, but my understanding of hires fix denoising strength is:
- normal SD goes from t=1.0 -> t=0.0 and tries to make the input (usually random noise) look like what you asked for
- img2img starts with an actual image instead of noise, so if we did the full 1.0 -> 0.0 process then it could look completely different
- denoising strength basically says 'pretend we've already done a bunch of steps', so denoise of 0.3 would go from t=0.3 -> t=0.0 instead of starting at the beginning
- before starting the actual sampler, there's code to inject noise into the image (`stochastic_encode`) like the previous step would have
- to inject this noise, we need to know the values from the model at the step before we start

and therein lies the bug... with 'uniform' scheduler/skip_type, 1.0 -> 0.0 basically goes like 1.0, 0.9, 0.8, etc., but with quadratic or logSNR it moves a bit faster at the start and a bit slower at the end, so like 1.0, 0.5, 0.25, 0.125, etc.

my stochastic_encode assumes a uniform distribution of timesteps, so if you match that then it should work for now.

still not sure why any of this would cause a tensor shape mismatch though :thinking:

**Comment by mart-hill** [comment]
My current samplers settings:
<img width="1679" alt="image" src="https://user-images.githubusercontent.com/55754730/235494596-46b4be22-131a-4132-bd28-d602da81e566.png">


**UniPC** with bh1 and time_uniform:
![00117-20230501190930_1954601011_sd-v15-ema-pruned](https://user-images.githubusercontent.com/55754730/235494483-23f5c60f-a736-4757-9715-fbf1f0f8148b.png)

**UniPC** with bh2 and time_uniform:
![00118-20230501191011_1954601011_sd-v15-ema-pruned](https://user-images.githubusercontent.com/55754730/235494719-759f53fe-bb3e-4a0b-92a8-2142d8ded65e.png)

**UniPC** with vary_coeff and time_uniform:
![00119-20230501191036_1954601011_sd-v15-ema-pruned](https://user-images.githubusercontent.com/55754730/235494756-698051bf-26c6-4ceb-b0c8-88d6572caa32.png)

In comparison, **Euler a**:
![00115-20230501190849_1954601011_sd-v15-ema-pruned](https://user-images.githubusercontent.com/55754730/235494850-78031de3-98e9-4ed6-9371-081f0d88ac51.png)

All the examples use:
```
starship in space in the orbit of a earth-like planet
Negative prompt: bad quality
Steps: 20, Sampler: UniPC (or Euler a), CFG scale: 7.5, Seed: 1954601011, Size: 512x768, Model hash: cc6cb27103 (sd-v15-ema-pruned.safetensor), Model: sd-v15-ema-pruned, VAE: blessed.vae, Clip skip: 2, ENSD: 31337

Time taken: 10.65sTorch active/reserved: 5142/6552 MiB, Sys VRAM: 5335/24576 MiB (21.71%)
```

The UniPC examples aren't bad here, but the one I presented up there used a model with noise offset embedded in, for better blacks/contrast, so it looks like this could trouble UniPC sampler somehow now.

**Comment by mart-hill** [comment]
...and yes, other "skip type" for UniPC do look worse. _Oh, again, this Generate freeze bug. Argh._

**Comment by mart-hill** [comment]
> what other unipc settings do you have? i've been sticking with 'skip_type' of 'uniform', I think other schedulers might make it look worse?
> 
> it's very strange that it works on that commit since it seems like the code shouldn't work (based on the `expand_dims` function being in the `bh` method but not where you're having a problem), and hasn't been changed recently.

Yes, I'll confirm, the image does look worse (without any negative prompt though):
![00104-20230501184459_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235510146-723db2c0-6488-412b-a0e4-4f4988700639.png)

Sorry for not noticing your request on time 🙂

**Comment by cool-bigdogs-tshirt** [comment]
what's the denoising strength on these? i got it working, but only for denoising of 0.5 :thinking: 
think i'm still picking the steps poorly

**Comment by mart-hill** [comment]
All those images weren't generated with hires fix. They should be untouched, so their "EXIF" also should be there, in case you'd like to drop them to "Process Image" tab. Just checked, they retained the generation data. 🙂

**Comment by mart-hill** [comment]
Uff, I hope I delivered all the data I could. I'm on, if there's need for more! 🙂

**Comment by vladmandic** [comment]
Bit off topic, but I find that UniPC needs but more steps to get to consistent image when running img2img or even tzt2img with Lora. Simple. Prompt and 10 is enough, for img2img, I'd start with min 20 and see how it behaves.

**Comment by mart-hill** [comment]
Yeah, the amount of steps is low, I didn't use any LoRAs though and those are txt2img, but I'll check with more steps.

**Comment by cool-bigdogs-tshirt** [comment]
anyone tried this in diffusers? because i'm pretty sure i got the code doing what it's doing in their implementation and it's just... not great. i don't think 'denoise strength' can really be implemented using timesteps the same way other samplers do it.

**Comment by mart-hill** [comment]
> Bit off topic, but I find that UniPC needs but more steps to get to consistent image when running img2img or even tzt2img with Lora. Simple. Prompt and 10 is enough, for img2img, I'd start with min 20 and see how it behaves.

Indeed, the UniPC sampler image with the increased amount of samples looks nicer, especially 25 steps:

20 steps:
![00046-20230502025430_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235556956-8b85fc21-c9a6-41d2-84cf-c3c5bc572a67.png)
Ugh.

25 steps:
![00047-20230502025517_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235556976-efb5e035-287d-4c9e-a9bd-55053e36eea2.png)
Way nicer! 

```
starship in space
Steps:20 and 25
Sampler: UniPC
CFG scale: 7
Seed: 1867324052
Size: 512x512
Model hash: 980122b4ec
Model: others_discordMixExperimental_v1
VAE: animefinal-full-pruned.vae
Clip skip: 2
ENSD: 31337
UniPC variant: vary_coeff
UniPC skip type: logSNR
```

Though, still **Euler a** looks a bit better with the exactly same parameters (save the sampler):

20 steps:
![00049-20230502025718_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235557379-242a9af6-6a76-424e-8ba2-75eea4ddc9f9.png)

25 steps:
![00048-20230502025655_1867324052_others_discordMixExperimental_v1](https://user-images.githubusercontent.com/55754730/235557380-efdfdaac-5944-49fd-8bf0-6a993c038734.png)

Most curios thing is, that the settings that errored on me with UniPC now work - I use xformers now, not SDP, though. Could that be the reason?

**Comment by vladmandic** [comment]
btw, anytime you see blobs with colorization mismatch, its a clear sign of sampler not done yet. same applies to pretty much all of them. your previous example is pretty much worst-case example of that:
https://user-images.githubusercontent.com/55754730/235510146-723db2c0-6488-412b-a0e4-4f4988700639.png

**Comment by cool-bigdogs-tshirt** [comment]
so this is one way to do it: https://i.imgur.com/9Mxdqwz.jpg
you ask for 20 steps with denoise of .2, it tells the scheduler to generate 100 timesteps (`20 // .2`) and runs the sampler on the last 20.

this is closer to how to how diffusers does it (and very close to the current implementation): https://i.imgur.com/Cv8Wre1.jpg
you ask for 20 steps with a denoise of .2, it generates only 20 timesteps, but starts them at .2 (instead of 1.0), and also runs for 20% of the steps (so only 4 steps with .2 denoising strength).

another option is this: https://i.imgur.com/ngkjklv.jpg
this always uses the full number of requested hires steps, and just varies the start of the schedule based on denoise strength. so 20 steps with denoise .2 is 20 steps of the schedule from 0.2 -> 0.0. i don't think i've seen img2img code that does this yet, so it feels a bit weird? also seems to be blurry no matter what

sooo none of these seem great. i'm going to be playing with other options, like using the second option, but keeping t_T at 0.5 and just using denoising strength to mess with the number of steps.

**Comment by cool-bigdogs-tshirt** [comment]
for more info on what i was tweaking, here's the last two rows of the far right column (.8 for quadratic and logsnr):
```
num_inference_steps 20, denoise_strength 0.7999999523162842, denoise_steps 20, init_timestep 0
tensor([1000,  905,  815,  730,  650], device='cuda:0', dtype=torch.int32)
latent timestep: tensor([0], device='cuda:0', dtype=torch.int32)
unipc called with steps 20, t_start 0.7999999523162842, t_end None
scheduling from t_0 0.001, t_T 0.7999999523162842
timesteps torch.Size([21]): tensor([0.8000, 0.7247, 0.6531, 0.5852, 0.5211, 0.4607, 0.4040, 0.3510, 0.3017,
        0.2562, 0.2144, 0.1763, 0.1419, 0.1113, 0.0844, 0.0612, 0.0417, 0.0259,
        0.0139, 0.0056, 0.0010], device='cuda:0'), new steps 20
```

```
num_inference_steps 20, denoise_strength 0.7999999523162842, denoise_steps 20, init_timestep 0
tensor([1000,  947,  890,  828,  760], device='cuda:0', dtype=torch.int32)
latent timestep: tensor([1], device='cuda:0', dtype=torch.int32)
unipc called with steps 20, t_start 0.7999999523162842, t_end None
scheduling from t_0 0.001, t_T 0.7999999523162842
timesteps torch.Size([21]): tensor([0.8000, 0.7419, 0.6791, 0.6112, 0.5387, 0.4625, 0.3849, 0.3090, 0.2384,
        0.1765, 0.1255, 0.0858, 0.0566, 0.0362, 0.0228, 0.0141, 0.0086, 0.0051,
        0.0033, 0.0022, 0.0010], device='cuda:0'), new steps 20
```

**Comment by vladmandic** [comment]
yup, those are good advanced tricks.
in theory, it should be possible to analyze interim results for compleness, use variable denoise factor and continue with as many steps as actually needed instead of having fixed number of steps or denoising. but that would require pretty much a new sampler (and it would need to be a single-pass, so likely euler based, cant do that with two pass dpm samplers nor with pre/post type samplers like unipc) :)
hey, maaaybe when i find some spare time? :)

**Comment by mart-hill** [comment]
> btw, anytime you see blobs with colorization mismatch, its a clear sign of sampler not done yet. same applies to pretty much all of them. your previous example is pretty much worst-case example of that: https://user-images.githubusercontent.com/55754730/235510146-723db2c0-6488-412b-a0e4-4f4988700639.png

Yet this was the exact image I've got with the settings I mentioned and which are in the image as well. Curious, right?

**Comment by vladmandic** [comment]
> Unless UniPC is meant to be used exclusively with hires fix?

not at all. unipc is primarily designed for txt2img. hires fix is internally img2img and that was enabled only recently.
but about all possible tunables for unipc, well, that's another story - i've never actually even tried `vary_coeff` :)

**Comment by mart-hill** [comment]
I almost exclusively use txt2img, and all those examples are from that tab. I'm yet to scratch the surface of img2img for more, than just inpainting a new eye or limb.  :)
I barely use batch generation, unless I want to see, if my training or mixing/training gave the results I'd like (then the batch of generated images would be mostly okay from the get go). 🙂

**Comment by vladmandic** [comment]
i understand that, what i'm saying is that as soon as you click on "hires fix", first pass will be txt2img, but second pass will internally be handled as img2img - and thats where you need extra steps to keep sampler happy. same with multidiffusion or any other such process.

**Comment by mart-hill** [comment]
Thanks for the explainings! It would seem that low amount of steps in UniPC won't make the sampler happy then. :)

**Comment by mart-hill** [comment]
After testing, I think [that model I've used](https://civitai.com/models/48087/discord-mix-experimental-garbage-mix) might require more samples, than usual, for generating an image, since it has noise offset data merged.

The following image has exactly the same "erroring" me earlier UniPC settings, and only 10 steps, but look:

![00089-20230502062245_1867324052_others_AOM2hard_hll4_final](https://user-images.githubusercontent.com/55754730/235580068-7510437c-5052-409b-9b41-a7cb35d40b25.png)

The only difference is the model, AOM2hard_hll4_final. :)

**Comment by vladmandic** [comment]
ok, so what's the current summary here? its a veeery long thread and issue is still marked as open :)

**Comment by mart-hill** [comment]
If this bug is no longer replicable to anyone else, I'll close it, is it okay?

**Comment by vladmandic** [comment]
if you're happy, i'm happy :)

**Comment by mart-hill** [comment]
Thank you so much for your hard work!

**Comment by cool-bigdogs-tshirt** [comment]
fyi <3
https://github.com/vladmandic/automatic/pull/721
